File: backend/app/core/cas.py
"""
CAS (Central Authentication Service) integration for IIITH.
Handles anonymous verification flow.
"""

import urllib.parse
from typing import Optional

import httpx

from app.core.config import settings


class CASClient:
    """CAS client for IIITH authentication."""

    def __init__(self):
        self.server_url = settings.CAS_SERVER_URL.rstrip('/')
        self.service_url = settings.CAS_SERVICE_URL

    def get_login_url(self, session_token: str) -> str:
        """
        Generate CAS login URL with session token as state parameter.

        The session token is passed as 'state' parameter to maintain
        verification session during CAS flow.
        """
        params = {
            'service': f"{self.service_url}?state={session_token}"
        }

        query_string = urllib.parse.urlencode(params)
        return f"{self.server_url}/login?{query_string}"

    async def validate_ticket(
            self, ticket: str, session_token: str
    ) -> Optional[str]:
        """
        Validate CAS ticket and return email if successful.

        Returns:
            Email address if validation successful, None otherwise.
        """
        validation_url = f"{self.server_url}/serviceValidate"

        params = {
            'ticket': ticket,
            'service': f"{self.service_url}?state={session_token}"
        }

        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(validation_url, params=params)
                response.raise_for_status()

                # Parse CAS XML response
                content = response.text

                # Simple XML parsing for CAS response
                if '<cas:authenticationSuccess>' in content:
                    # Extract username (email) from response
                    start = content.find('<cas:user>') + len('<cas:user>')
                    end = content.find('</cas:user>')

                    if start > len('<cas:user>') - 1 and end > start:
                        email = content[start:end].strip()

                        # Validate IIITH email format
                        if self._is_valid_iiith_email(email):
                            return email

                return None

        except Exception as e:
            print(f"CAS validation error: {e}")
            return None

    def _is_valid_iiith_email(self, email: str) -> bool:
        """Validate that email is from IIITH domain."""
        valid_domains = [
            '@students.iiit.ac.in',
            '@research.iiit.ac.in',
            '@iiit.ac.in'
        ]

        email_lower = email.lower()
        return any(email_lower.endswith(domain) for domain in valid_domains)


# Global CAS client instance
cas_client = CASClient()

File: frontend/src/lib/cas.ts
/**
 * CAS (Central Authentication Service) client for IIITH verification
 * Handles anonymous verification flow
 */

const CAS_SERVER_URL = process.env.CAS_SERVER_URL || 'https://login.iiit.ac.in/cas'
const CAS_SERVICE_URL = process.env.CAS_SERVICE_URL!
const ALLOWED_EMAIL_DOMAINS = (process.env.ALLOWED_EMAIL_DOMAINS || 'students.iiit.ac.in,research.iiit.ac.in,iiit.ac.in').split(',')

/**
 * Generate CAS login URL with session token as state parameter
 * The session token is passed as 'state' parameter to maintain
 * verification session during CAS flow.
 */
export function getCASLoginUrl(sessionToken: string): string {
    const serviceUrl = `${CAS_SERVICE_URL}?state=${encodeURIComponent(sessionToken)}`
    const params = new URLSearchParams({ service: serviceUrl })
    return `${CAS_SERVER_URL.replace(/\/$/, '')}/login?${params.toString()}`
}

/**
 * Validate CAS ticket and return email if successful
 * Returns email address if validation successful, null otherwise
 */
export async function validateCASTicket(
    ticket: string,
    sessionToken: string
): Promise<string | null> {
    const validationUrl = `${CAS_SERVER_URL.replace(/\/$/, '')}/serviceValidate`
    const serviceUrl = `${CAS_SERVICE_URL}?state=${encodeURIComponent(sessionToken)}`

    try {
        const response = await fetch(
            `${validationUrl}?ticket=${encodeURIComponent(ticket)}&service=${encodeURIComponent(serviceUrl)}`,
            { method: 'GET' }
        )

        if (!response.ok) {
            console.error('CAS validation HTTP error:', response.status)
            return null
        }

        const content = await response.text()

        // Parse CAS XML response
        if (content.includes('<cas:authenticationSuccess>')) {
            // Extract username (email) from response
            const startTag = '<cas:user>'
            const endTag = '</cas:user>'
            const startIndex = content.indexOf(startTag)
            const endIndex = content.indexOf(endTag)

            if (startIndex !== -1 && endIndex > startIndex) {
                const email = content.substring(startIndex + startTag.length, endIndex).trim()

                // Validate IIITH email format
                if (isValidIIITHEmail(email)) {
                    return email
                }
            }
        }

        return null
    } catch (error) {
        console.error('CAS validation error:', error)
        return null
    }
}

/**
 * Check if email is from a valid IIITH domain
 */
export function isValidIIITHEmail(email: string): boolean {
    const emailLower = email.toLowerCase()
    return ALLOWED_EMAIL_DOMAINS.some(domain =>
        emailLower.endsWith(`@${domain.trim()}`)
    )
}

/**
 * Generate a secure random session token
 */
export function generateSessionToken(): string {
    const array = new Uint8Array(32)
    crypto.getRandomValues(array)
    return Array.from(array, byte => byte.toString(16).padStart(2, '0')).join('')
}

/**
 * Get verification session expiration time
 */
export function getSessionExpirationMinutes(): number {
    return parseInt(process.env.VERIFICATION_SESSION_EXPIRE_MINUTES || '30', 10)
}
